pyogrio/_io.pyx

```diff
@@ -2178,3 +2180,182 @@ def ogr_write(
+    str encoding,
+    object dataset_kwargs,
+    object layer_kwargs,
+    bint append=False,
+    dataset_metadata=None,
+    layer_metadata=None,
+):
+    IF CTE_GDAL_VERSION < (3, 8, 0):
+        raise RuntimeError("Need GDAL>=3.8 for Arrow write support")
+
+    cdef OGRDataSourceH ogr_dataset = NULL
+    cdef OGRLayerH ogr_layer = NULL
+
+    if not encoding:
+        encoding = locale.getpreferredencoding()
+
+    layer_created = create_ogr_dataset_layer(
+        path, layer, driver, crs, geometry_type, encoding,
+        dataset_kwargs, layer_kwargs, append,
+        dataset_metadata, layer_metadata,
+        &ogr_dataset, &ogr_layer,
+    )
+    if geometry_name:
+        opts = {"GEOMETRY_NAME": geometry_name}
+    else:
+        opts = {}
+    cdef char **options = dict_to_options(opts)
+
+    stream_capsule = arrow_obj.__arrow_c_stream__()
+    write_arrow_stream_capsule(ogr_layer, stream_capsule, geometry_name, options)
+
+    ### Final cleanup
+    if ogr_dataset != NULL:
+        GDALClose(ogr_dataset)
+
+        # GDAL will set an error if there was an error writing the data source
+        # on close
+        exc = exc_check()
+        if exc:
+            raise DataSourceError(f"Failed to write features to dataset {path}; {exc}")
+
+
+IF CTE_GDAL_VERSION >= (3, 8, 0):
+
+    cdef write_arrow_stream_capsule(OGRLayerH destLayer, object capsule, str geometry_name, char** options = NULL):
+        cdef ArrowSchema schema
+        cdef ArrowArray array
+
+        cdef ArrowArrayStream* stream = <ArrowArrayStream*>PyCapsule_GetPointer(
+            capsule, "arrow_array_stream"
+        )
+        if stream == NULL:
+            raise RuntimeError("No valid stream.")
+
+        if stream.release == NULL:
+            raise RuntimeError("Arrow Array Stream was already released.")
+
+        errcode = stream.get_schema(stream, &schema)
+        if errcode != 0:
+            stream.release(stream)
+            raise RuntimeError("Error while accessing schema from stream.")
+
+        try:
+            create_fields_from_arrow_schema(destLayer, &schema, options, geometry_name)
+        except Exception as e:
+            schema.release(&schema)
+            stream.release(stream)
+            raise e
+
+        while True:
+            errcode = stream.get_next(stream, &array)
+            if errcode != 0:
+                schema.release(&schema)
+                stream.release(stream)
+                raise RuntimeError("Error while accessing batch from stream.")
+
+            # We've reached the end of the stream
+            if array.release == NULL:
+                break
+
+            errcode = OGR_L_WriteArrowBatch(destLayer, &schema, &array, options)
+            if not errcode:
+                if array.release != NULL:
+                    array.release(&array)
+
+                schema.release(&schema)
+                stream.release(stream)
+                raise RuntimeError("Error while writing batch to OGR layer.")
```

Comment:
@himikof good point. It would be good to come up with a test case that would trigger an error, to test this (some unsupported data type for a certain driver?), because it's not super clear to me if GDAL actually sets an error in that case (since this function returns a bool and not OGRErr (like other OGR functions where we check the set exception)

Reply:
Found an example with FGB (first tried shapefile, but it seems GDAL just converts anything Shapefile doesn't support into its string representation and write it anyway ..):

```
In [1]: import shapely

In [2]: import pyogrio

In [3]: table = pa.table({"geometry": shapely.to_wkb(shapely.from_wkt(["POINT (0 0)", "POINT (1 1)"])), "col": [[0,1], [2,3,4]]})

In [4]: pyogrio.raw.write_arrow(table, "test_list.fgb", geometry_type="Point", geometry_name="geometry")
...
CPLE_AppDefinedError: ICreateFeature: Missing implementation for OGRFieldType 13
```

So with checking `exc_check()`, we indeed get a more targetted error message (it's not a super clear error message TBH, but that's something to improve on the GDAL side ;))